### ISSUE TITLE
Create Kubernetes connection pool

### ISSUE THREAD

**Issue by jacobowitz** [jina-ai/jina#3272]
We need a class to manage connections (for either grpc or zmq) in Kubernetes

**Comment by jacobowitz** [comment]
draft pr for the pool used in benchmark: https://github.com/jina-ai/jina/pull/3275
draft pr for benchmarking code: https://github.com/jina-ai/jina/pull/3370


result

```
5k requests
flow with 4 hops
Client running in same datacenter

proxy
total time: 97.56727075576782 s
per req: 19.513454151153564 m
per req g2g: 13.448080749999999 ms

connection pool
total time end2end: 70.56362891197205 s
per req end2end: 14.11272578239441 ms
per req g2g: 8.8882288 ms
```

```
long flow (+4 executors chained)

proxy
total time: 156.72174072265625 s
per req: 31.34434814453125 ms
request avg g2g: 24.813048 ms

pool
total time: 108.6645860671997 s
per req: 21.73291721343994 ms
request avg g2g: 15.978156 ms
```


So I think we have to options:

**1. Use proxy based load balancing:**

- Adds overhead of ~ 1.1 ms per executor in flow (relative to pool)
- No need for logic on our side
- Users need to deploy third party service mesh (istio)

**2. Use  custom connection pool implementation:**

- Slightly faster
- Existing codes needs some polishing
- No need for 3rd party tools, but more code to maintain for us

**Comment by jacobowitz** [comment]
We decided to go with both options

**Comment by jacobowitz** [comment]
PR: https://github.com/jina-ai/jina/pull/3275

**Comment by JoanFM** [comment]
Closed by #3275
